### PR TITLE
Create paragraph level formatting toolbar

### LINKED ISSUE #6501: Rich Editor - Paragraph Level Formatting toolbar
Implement the UI of the paragraph level toolbar. This should always be on the beginning of the currently selected line.

It's buttons should have the following options:

- H1
- H2
- code-block
- blockquote

### LINKED ISSUE #6572: Rich Editor - Fire event on "escape"
Each react component in the editor should use this event to close as needed when escape is hit.

### PR DESCRIPTION
**Pull request by Charrondev**
Closes #6501, Closes #6572 

Creates a paragraph level formatting toolbar for the editor. Also ensures that all existing flyouts close each other by firing an event.

I needed to make some updates to the BlockQuote, CodeBlock, and Spoiler blot so that they could interchange with each other. I've made a new generic blot for the content, and a helper function to generate their wrapper Blot.

There are still a few things to wrap up in these blots, but they aren't really relevant to this PR.

### CODE REVIEW

**Comment by Charrondev** [comment]
I'm putting a WIP on this because I still have a couple of things to resolve. I have something to fix in the tests, and I need to finish positioning the nub for the paragraph toolbar.

File: applications/dashboard/scss/new-features/_spoiler.scss [hunk 1]
```diff
@@ -39,7 +39,7 @@ $buttonSpoiler: (
     background-color: $spoiler-bg;
 
     .spoiler-content {
-        display: none;
+        //display: none;
```
Comment (slafleche) [inline applications/dashboard/scss/new-features/_spoiler.scss]: Delete if not being used

File: plugins/rich-editor/src/scripts/components/Icons.jsx [hunk 2]
```diff
@@ -66,3 +66,46 @@ export function close() {
         <path fill="currentColor" d="M12,10.6293581 L5.49002397,4.11938207 C5.30046135,3.92981944 4.95620859,3.96673045 4.69799105,4.22494799 L4.22494799,4.69799105 C3.97708292,4.94585613 3.92537154,5.29601344 4.11938207,5.49002397 L10.6293581,12 L4.11938207,18.509976 C3.92981944,18.6995387 3.96673045,19.0437914 4.22494799,19.3020089 L4.69799105,19.775052 C4.94585613,20.0229171 5.29601344,20.0746285 5.49002397,19.8806179 L12,13.3706419 L18.509976,19.8806179 C18.6995387,20.0701806 19.0437914,20.0332695 19.3020089,19.775052 L19.775052,19.3020089 C20.0229171,19.0541439 20.0746285,18.7039866 19.8806179,18.509976 L13.3706419,12 L19.8806179,5.49002397 C20.0701806,5.30046135 20.0332695,4.95620859 19.775052,4.69799105 L19.3020089,4.22494799 C19.0541439,3.97708292 18.7039866,3.92537154 18.509976,4.11938207 L12,10.6293581 Z"/>
     </svg>;
 }
+
+export function pilcro() {
```
Comment (slafleche) [inline plugins/rich-editor/src/scripts/components/Icons.jsx]: "Pilcrow" not "pilcro"

File: plugins/rich-editor/src/scripts/components/ParagraphEditorToolbar.jsx [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+ * @copyright 2009-2018 Vanilla Forums Inc.
+ * @license https://opensource.org/licenses/GPL-2.0 GPL-2.0
+ */
+
+import React from "react";
+import * as PropTypes from "prop-types";
+import Quill from "quill/core";
+import Emitter from "quill/core/emitter";
+import EditorToolbar from "./EditorToolbar";
+import { pilcro as PilcroIcon } from "./Icons";
```
Comment (slafleche) [inline plugins/rich-editor/src/scripts/components/ParagraphEditorToolbar.jsx]: pilcrow

File: plugins/rich-editor/src/scripts/components/ParagraphEditorToolbar.jsx [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        }
+
+        if (source !== Quill.sources.SILENT) {
+            this.setState({
+                showMenu: false,
+            });
+        }
+    };
+
+    getPilcroStyles() {
```
Comment (slafleche) [inline plugins/rich-editor/src/scripts/components/ParagraphEditorToolbar.jsx]: pilcrow

File: plugins/rich-editor/src/scripts/components/ParagraphEditorToolbar.jsx [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+            visibility: "hidden",
+            position: "absolute",
+            zIndex: -1,
+        };
+
+        return this.state.showMenu ? {} : hiddenStyles;
+    }
+
+    /**
+     * Click handler for the Pilcro
```
Comment (slafleche) [inline plugins/rich-editor/src/scripts/components/ParagraphEditorToolbar.jsx]: pilcrow

**Comment by slafleche** [comment]
Bug:
1. Make 3 lines
2. Make 2nd line an "h2"
3. click on 3rd line
4. Open paragraph menu, the "h2" is active

**Comment by Charrondev** [comment]
> Selecting code on selected part of a heading doesn't convert it into a separate paragrph

I've opened a proposal to address this. The way we do may need to more general than just this one case.
https://github.com/vanilla/roadmap/pull/29

I've resolved the rest of the feedback as well. That included removing the code block wrapper.

**Pull request merged**